language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -28,8 +34,15 @@ import           Language.PlutusCore.Type
 import           Lens.Micro
 import           PlutusPrelude
 
-newtype TypeConfig = TypeConfig
-    { _typeConfigNormalize :: Bool  -- ^ Whether we normalize type annotations
+-- | Mapping from 'DynamicBuiltinName's to their 'Type's.
+newtype DynamicBuiltinNameTypes = DynamicBuiltinNameTypes
+    { unDynamicBuiltinNameTypes :: Map DynamicBuiltinName (Quote (Type TyName ()))
```

Comment:
Why can't we just have a `Map DynamicBuiltinName (Type TyName ())` which we clone? Also, is there some reason this is `Type TyName ()` instead of `Type TyNameWithKind ()`?

Reply:
> Why can't we just have a `Map DynamicBuiltinName (Type TyName ())` which we clone? 

Types of built-ins are closed, so we just generate them unique and no cloning is required. It's simpler than using cloning which is specifically designed for things that are not closed.

> Also, is there some reason this is `Type TyName ()` instead of `Type TyNameWithKind ()`

Yes, to match static built-in names. The pipeline is: `TypeScheme` -> `Type TyName` -> `Type TyNameWithKind` with the last transformation happening inside the type checker for both static and dynamic built-ins.

However it's probably indeed more efficient to precompute kinds. I'll play with this.

Reply:
It's `Type TyNameWithKind ()` now.